[Go] preliminary support for Go 1.18

mitranim opened the pull request:
Features:

  * Support for `~`.
  * Support for `any` and `comparable`.
  * Support for type unions.
  * Support for general interfaces (with type unions).
  * Support for type parameters in generic functions.
  * Support for type parameters in method receivers of generic types.
  * Support for type arguments following type names.
  * Support for type parameters in type declarations.
  * Support for embedded parametrized types in struct definitions.
  * Support for type arguments between function name and function arguments.
  * Support for `invalid.illegal` for dangling `)]}`.
  * Slightly better meta scopes in type definitions.

Known regressions:

  * Embedded interfaces are no longer scoped as "inherited class", due to ambiguities.
  * "Namespace" idents such as `ident.` no longer support a comment between identifier and dot.
  * Last ident in parens followed by opening paren is no longer scoped as a function call, due to false positives and ambiguities.

Known issues and limitations:

  * Detection of type parameter lists and type argument lists is incomplete. In some places it relies on regex lookahead, which doesn't work across multiple lines. There are also edge case ambiguities which are supported by the Go compiler but not by our syntax.

Misc tweaks in comments:

  * Convert anon to named contexts for compatibility with overrides.
  * Avoid false positives when scoping javadoc bullets.
  * Use `prototype` to reduce repetition.

Closes #3281.

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -175,37 +214,45 @@ contexts:
-        - match: ^\s*(\*)(?!/)
-          captures:
-            1: punctuation.definition.comment.go
+  pop-comment-line-annotation:
+    - meta_scope: meta.annotation.go comment.line.go
+    - match: \S+
+      scope: meta.variable.parameter.go
+    # End the annotation scope at EOL, but stretch the comment scope
+    # indefinitely to the right.
+    - match: $
+      set: pop-comment-line-normal
```
This context seems to be unused and may be removed.

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -175,37 +214,45 @@ contexts:
+    # Indentation, followed by exactly one space, followed by *.
+    # This enables support for javadoc comments, while avoiding false
+    # positives in non-javadoc comments that use "*" as simple list bullets.
+    - match: ^(?:\t|[ ]{2}|[ ]{4}|[ ]{8})*[ ](\*)(?!/)
```
Smart!

deathaxe commented on Go/syntax_test_go.go:
```diff
@@ -35,22 +34,30 @@ You may have to disable Go-specific linters when working on this file.
 //                            ^^^^^^^^^^^ comment.line.go
 
     /*
+//  ^^ comment.block.go punctuation.definition.comment.begin.go
 // ^ -comment
```
We should probably avoid duplicated tests (of `comment.block). Instead we could ensure correct boundaries by:

```suggestion
// ^ -comment
//  ^^ comment.block.go punctuation.definition.comment.begin.go
//    ^^ comment.block.go - punctuation
```

deathaxe commented on Go/syntax_test_go.go:
```diff
@@ -35,22 +34,30 @@ You may have to disable Go-specific linters when working on this file.
 //  ^^^^^^^^ comment.block.go
     */
-//  ^^ comment.block.go
+//  ^^ comment.block.go punctuation.definition.comment.end.go
```
The counter part for checking boundaries of the previous suggestion.

```suggestion
//^^ comment.block.go - punctuation
//  ^^ comment.block.go punctuation.definition.comment.end.go
```

deathaxe commented on Go/syntax_test_go.go:
```diff
@@ -35,22 +34,30 @@ You may have to disable Go-specific linters when working on this file.
 
     /* * */
 // ^ -comment
+//     ^ -punctuation.definition.comment.go
 //  ^^^^^^^ comment.block.go
```
Just a tiny nit pick.

```suggestion
//  ^^^^^^^ comment.block.go
//     ^ -punctuation.definition.comment.go
```

jrappen commented:
* It might make sense to move tests to `./Go/tests/` and split them up.
* You're still using syntax v1. Was there any reason not make use of newer v2 syntax engine features?

mitranim commented:
> * It might make sense to move tests to `./Go/tests/` and split them up.

I have no strong opinion. It has obvious upsides but also some downsides. Search-and-replace opens N files, split-panel becomes slightly more annoying to use.

> * You're still using syntax v1. Was there any reason not make use of newer v2 syntax engine features?

No reason. Updated to v2.

deathaxe commented:
FWIW There's no hard restriction with regards to used syntax features and syntax version. V2 just gates some backward incompatible changes to the syntax engine, which require a syntax to opt-in. Most of them are related to how meta scopes are applied in push/set/pop scenarios.

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -70,19 +71,23 @@ variables:
   # functions and types.
   ident: \b(?!{{keyword}})[[:alpha:]_][[:alnum:]_]*\b
 
+  keyword_or_ident: \b[[:alpha:]_][[:alnum:]_]*\b
```
May probably want to use it in `ident`?

```yaml
ident: (?!{{keyword}}){{keyword_or_ident}}
```

mitranim replied:
Wasn't sure if this would work properly because unlike the previous version, this inserts `\b` into the middle of the regex. But tests pass so you might be right.

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -296,18 +334,23 @@ contexts:
       scope: variable.function.go
 
   # See notes on `match-call-or-type-conversion`.
   pop-call-or-type-conversion:
+    - include: pop-identifier-blank
```
It appears the whole context is unused.

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -388,49 +455,71 @@ contexts:
+
+  pop-type-assertion:
+    - include: pop-on-paren-end
+    - match: (?=\S)
+      push: pop-type
```
```suggestion
    - include: match-type
```

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -967,175 +1059,269 @@ contexts:
       pop: true
 
+  match-type:
+    - match: (?=\S)
+      push: pop-type
```
I'd suggest the following to remove one extra context switch. I think this trivial kind of duplication doesn't hurt.

```suggestion
    - match: (?=\S)
      push: [pop-after-type, pop-type-single]
```

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -873,67 +944,89 @@ contexts:
+    - include: pop-before-terminator
+    - match: (?=\()
+      set: pop-func-parameter-list
+    - match: (?=\S)
+      set: pop-type
```
The way `pop-type` is implemented the following would work, too, and avoid one extra context switch.

```suggestion
    - include: pop-type
```

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -967,175 +1059,269 @@ contexts:
     - include: match-comma
     - include: match-ellipsis
+    - include: match-keywords
     - match: (?=\S)
       push: pop-type
```
The way `match-type` is implemented the following would work, too, and avoid one extra context switch.

```suggestion
    - include: match-type
```

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -967,175 +1059,269 @@ contexts:
+    - include: match-any
+
+  pop-struct-field-meta:
+    - include: pop-before-brace-end
+    - include: pop-before-terminator
+    - include: pop-type
+    - include: match-any
```
I have a feeling about `pop-type` to need to be the last line as it starts with `(?=\S)`, but I may be wrong.

```suggestion
    - include: match-any
    - include: pop-before-brace-end
    - include: pop-before-terminator
    - include: pop-type
```

mitranim replied:
The "pop-before" rules must come first, but you're right in that `match-any` wouldn't work after `pop-type`, so I'll just remove `match-any` from here. It was included for graceful degradation, like in some other places.

deathaxe reviewed (commented):
Here's another set of suggestions.

jrappen commented:
Appreciate the effort to name contexts.

TheSecEng commented:
Is this also going to support syntax based code folding ?

jrappen commented:
@TheSecEng Currently no changes for that, yet.

@mitranim For reference the [`master`](https://github.com/sublimehq/Packages/commits/master) branch already has some commits for the new code folding stuff in other languages.

jrappen commented:
Maybe address #3228 as well with this PR?

deathaxe commented:
Syntax based code folding should work pretty well, once each pushed context receives a meta scope so neiboured parens can be detected as separate folding begin/end markers.

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -799,37 +859,31 @@ contexts:
+    - match: \(
+      scope: punctuation.section.parens.begin.go
+      set: pop-type-declaration-list
+    - match: (?=\S)
+      set: pop-type-declaration
```
Maybe 

```suggestion
    - include: pop-type-declaration
```

deathaxe commented on Go/Go.sublime-syntax:
```diff
@@ -1202,72 +1352,83 @@ contexts:
+        1: variable.other.go
+        2: punctuation.accessor.dot.go
+
+  meta-function-declaration:
+    - meta_scope: meta.function.declaration.go
```
All those meta-only contexts should contain `- meta_include_prototype: false` to avoid `prototype`s from being prepended. Even if the language itself doesn't define any `prototype` context any inherited syntax might do so in order to inject templating patterns etc.

```suggestion
  meta-function-declaration:
    - meta_include_prototype: false
    - meta_scope: meta.function.declaration.go
```

deathaxe reviewed (commented):
The linked issue still contains some open tasks. 

Do you plan to integrate them in this PR?

In case those are planned for future PRs, would you mind adding some small reminder issues with existing limitations and possible enhancements to certain code constructs - maybe including small examples?

mitranim commented:
Apologies for slow responses. The linked issue had multiple progress reports which were all out of dated, and it wasn't meant to be a progress status of _this_ PR. However, there are other known defects. I'm building a list now.

mitranim commented:
Known defects / missing features:

* [ ] In parameter lists, types with type arguments in square brackets (inline type instantiation) should be treated as types, rather than regular parameters.

```go
interface{
  Method(Type[TypeArg])
}

func(Type, Type[TypeArg])
```

* [ ] In embedded interfaces with type arguments, type arguments should be scoped as `variable.other.type` rather than `variable.other`.

```go
interface{
  EmbeddedInterface[TypeArg]
}
```

* [ ] "Function call" detection should work with type arguments of arbitrary complexity, rather than just with identifiers.

```go
FuncName[Type[TypeArg]]()
```

* [ ] Parameter lists should support types of arbitrary complexity such as `[][]Ident`.

```go
func FuncName(param [][]Type)
```

deathaxe reviewed (dismissed):
This PR provides many improvements with regards to type support.

Not supporting block quotes everywhere in fully-qualified identifiers is acceptable. Many other syntaxes don't do so at all, even though valid by language specs.

Work on Java syntax teached me required complexity needed to handle all those typing and generics "perfect" especially with regards to multi-line support.

So I think we could add this PR as improvement and add an issue with the open "TODOs" to come back to them later. Maybe fixing them in smaller pieces?

mitranim commented:
I've been using this implementation for months, and while it's still incomplete, it's a massive improvement over not having support for Go 1.18. I would suggest merging this sooner, to improve user experience in the next stable build, and then I'll open a new issue for the remaining unresolved problems.

deathaxe commented:
Could you probably have a look at the conflicts?

mitranim commented:
> Maybe you could add to the commit message (the opening comment) that you're addressing RFC issue 2569 with the scopes for doc blocks.

This confuses me. The issue seems to be about syntax highlighting inside docblocks, for stuff like `@param{type}`. Our Go syntax does not support this and hopefully never will, because Go generates documentation from source code, making Javadoc unnecessary. Not sure if this RFC applies.

mitranim commented:
For `set: [one, two]` that easily fits under 80 char, I prefer a one-liner, but that's very subjective. I wouldn't mutiny if someone changed it later. Not seeing a strong convention and don't have a strong opinion.

For long regexes that are lists of words such as `(?:one|two|three)`, I subjectively prefer a one-liner for code compactness. They tend to appear in `variables`, and you can easily scroll past a one-liner, while a multi-liner requires you to "parse" code. However, a flat multi-line structure makes Git diffs more useful:

```diff
  predeclared_type: |-
    \b(?x:
-     bool
+     any
+     |bool
      |byte
+     |comparable
      |complex64
      |complex128
      ...
    )\b
```

...as opposed to:

```diff
-  predeclared_type: \b(?:byte|complex64|complex128| ...
+  predeclared_type: \b(?:any|bool|byte|comparable|complex64|complex128 ...
```

So if we're changing variable lists, it should be by flattening them vertically, IMO.

deathaxe commented:
> For set: `[one, two]` that easily fits under 80 char

I agree, one-liners beeing well suited in case they fit into 80 chars boundary.

Multi-line form got a bit more popular just because of many contexts being pushed at once in syntaxes like Java or JavaScript. They may produce more readable code, if context names get longer.

I've tried to keep with one style recently, but there's no strong convention about it, nor a need for one.

> a flat multi-line structure makes Git diffs more useful

Agree. With a look at Ruby, which makes already use of it, and some experiments with applying it to PHP, I'd be a bit concerned about the huge amount of lines added - especially for syntaxes with insane long lists of builtin funcs/vars/... (such as PHP).

TheSecEng commented:
Been using this for a month or so. Great work!

michaelblyons commented:
@patrickrgaffney @jakelucas Any comments?

jakelucas commented:
> @patrickrgaffney @jakelucas Any comments?

I can only assume I've been pinged here by mistake
Looks good to me
